flatliners/clusteralertcorr.py

```diff
@@ -0,0 +1,132 @@
+import math
+import pandas as pd
+from datetime import datetime
+from .baseflatliner import BaseFlatliner
+
+class ClusterAlertCorrelation(BaseFlatliner):
+
+    '''
+    This class contains the code for correlation of alerts for every cluster
+
+    '''
+
+    def __init__(self):
+        print(" Inside alert corr")
+        super().__init__()
+        ## Hold the values for different clusters
+        self.clusters = dict()
+        #TODO : save correlation over time
+        self.clusters_correlation = dict()
+        ### To keep track of previous publish
+        self.timestamp = datetime.fromtimestamp(0)
+
+    def on_next(self, x):
+        """ On each data loading it will create dataframe and then pivot frame
+        """
+        # print(f"got {x}")
+
+        if not self.metric_name(x) == 'alerts':
+            return
+
+        alert_name = self.metric_label(x, 'alertname')
+        cluster_id = self.cluster_id(x)
+
+        if cluster_id not in self.clusters:
+            self.clusters[cluster_id] = pd.DataFrame(columns=['_id', 'timestamp', 'alertname'])
+            self.clusters_correlation[cluster_id] = pd.DataFrame()
+
+        previous_df = self.clusters[cluster_id]
+
+        ### Dataframe creation
+        timestamp = []
+        cluster = []
+        alert_name_list = []
+
+        for elem in self.metric_values(x):
+            timestamp.append(datetime.fromtimestamp(elem[0]))
+            cluster.append(cluster_id)
+            alert_name_list.append(alert_name)
+        df = previous_df.append\
+            (pd.DataFrame({'_id': cluster, 'timestamp': timestamp, 'alertname': alert_name_list}))
+        df.reset_index(drop= True, inplace= True)
+
+        self.clusters[cluster_id] = df
+
+
+        ### Publishing for every hour
+        if(datetime.timestamp(timestamp[-1]) - datetime.timestamp(self.timestamp)) > 3600:
+            self.print_values(self.clusters, timestamp[-1])
+            self.timestamp = timestamp[-1]
+            self.publish(self.clusters)
+            for clust_id in self.clusters.keys():
+                count_frame = self.count_alert_per_cluster(self.clusters[clust_id])
+                self.print_corr(count_frame, clust_id)
+                self.publish(count_frame)
```

Comment:
You should not `publish` multiple things. If you need the `counts` for the `correlation` then either create a new flatliner (if you think somebody else might be interested in them) or just store them.
But one flatliner should publish only **one** type of things

Reply:
Did not know. will correct.

Reply:
Done!